synapse/storage/databases/main/roommember.py

```diff
@@ -187,27 +187,48 @@ def _check_safe_current_state_events_membership_updated_txn(
 
     @cached(max_entries=100000, iterable=True)
     async def get_users_in_room(self, room_id: str) -> List[str]:
+        """
+        Returns a list of users in the room sorted by longest in the room first
+        (aka. with the lowest depth). This is done to match the sort in
+        `get_current_hosts_in_room()` and so we can re-use the cache but it's
+        not horrible to have here either.
+        """
+
         return await self.db_pool.runInteraction(
             "get_users_in_room", self.get_users_in_room_txn, room_id
         )
 
     def get_users_in_room_txn(self, txn: LoggingTransaction, room_id: str) -> List[str]:
+        """
+        Returns a list of users in the room sorted by longest in the room first
+        (aka. with the lowest depth). This is done to match the sort in
+        `get_current_hosts_in_room()` and so we can re-use the cache but it's
+        not horrible to have here either.
+        """
         # If we can assume current_state_events.membership is up to date
         # then we can avoid a join, which is a Very Good Thing given how
         # frequently this function gets called.
         if self._current_state_events_membership_up_to_date:
             sql = """
-                SELECT state_key FROM current_state_events
-                WHERE type = 'm.room.member' AND room_id = ? AND membership = ?
+                SELECT c.state_key FROM current_state_events as c
+                /* Get the depth of the event from the events table */
+                INNER JOIN events AS e USING (event_id)
```

Comment:
Time-wise, I think by the nature of the extra complexity, it will take more time but it's not something I optimized for here. If we don't like this trade-off, we could remove the cache-shortcut lookup in `get_current_hosts_in_room` around it.

Is there another concern about the order itself?

Reply:
fwiw, here are some first-run timings. Not sure how much the first one affects the second one 🤷 

```
$ psql synapse
synapse=# \timing on

synapse=# SELECT c.state_key FROM current_state_events as c
/* Get the depth of the event from the events table */
INNER JOIN events AS e USING (event_id)
WHERE c.type = 'm.room.member' AND c.room_id = '!OGEhHVWSdvArJzumhm:matrix.org' AND membership = 'join'
/* Sorted by lowest depth first */
ORDER BY e.depth ASC;

Time: 4374.227 ms (00:04.374)


synapse=# SELECT c.state_key FROM current_state_events as c
WHERE c.type = 'm.room.member' AND c.room_id = '!OGEhHVWSdvArJzumhm:matrix.org' AND membership = 'join';

Time: 1947.409 ms (00:01.947)
```

Reply:
Only doubling isn't as bad as I was expecting, so maybe it's fine. Possibly worth checking a graph for how much time is spent on `get_users_in_rooms` and thinking if it would burn us if that time were doubled.

Reply:
I've looked at the usages and there are some areas where it doesn't feel great to increase the time (calculating if an appservice is interested in event (this could probably be smarter though anyway), calculating `join_authorised_via_users_server` in a `/make_join` (idk how important this is), newly joined rooms in `/sync`). Overall seems kinda fine ⏩

I can create separate PRs to fix up the following areas first, then we can merge this after those places are changed.

There are a few areas where it feels like we shouldn't be using `get_users_in_room` and can just use a direct lookup for the single user:

 - https://github.com/matrix-org/synapse/blob/a25a37002c851ef419d12925a11dd8bf2233470e/synapse/handlers/events.py#L176-L177
 - https://github.com/matrix-org/synapse/blob/a25a37002c851ef419d12925a11dd8bf2233470e/synapse/handlers/room.py#L1287-L1288
  - https://github.com/matrix-org/synapse/blob/a25a37002c851ef419d12925a11dd8bf2233470e/synapse/handlers/message.py#L764-L765
 - https://github.com/matrix-org/synapse/blob/a25a37002c851ef419d12925a11dd8bf2233470e/synapse/handlers/room_member.py#L1623-L1624 
 - https://github.com/matrix-org/synapse/blob/a25a37002c851ef419d12925a11dd8bf2233470e/synapse/server_notices/server_notices_manager.py#L115

 There are a few places where we should use `get_current_hosts_in_room`:
 
  - https://github.com/matrix-org/synapse/blob/a25a37002c851ef419d12925a11dd8bf2233470e/synapse/handlers/directory.py#L86-L87
  - https://github.com/matrix-org/synapse/blob/a25a37002c851ef419d12925a11dd8bf2233470e/synapse/handlers/directory.py#L290-L291
  - https://github.com/matrix-org/synapse/blob/a25a37002c851ef419d12925a11dd8bf2233470e/synapse/handlers/presence.py#L2054-L2055
  - https://github.com/matrix-org/synapse/blob/a25a37002c851ef419d12925a11dd8bf2233470e/synapse/handlers/typing.py#L365-L366

Reply:
Addressed these mis-uses in:

 - https://github.com/matrix-org/synapse/pull/13605
 - https://github.com/matrix-org/synapse/pull/13608

Reply:
I'd expect `get_users_in_room` to be visible in the Database section of the Grafana dashboard, but it seems like it might be so insignificant that it doesn't show up there. I think that probably hints at the answer.

Reply:
Seems mostly at `0.5%` of total time, sometimes `2%`. But it's down in the noise of the rest of the txn's we do if you show all of them again.

Doubling 3ms - 15ms average time isn't horrible but would be good to check in on this after this PR merges.

![](https://user-images.githubusercontent.com/558581/186487406-eff48d16-2e07-442b-921a-1c1659b7a78f.png)

Reply:
Here is some follow-up. I think this PR was rolled out on `matrix.org` (or maybe just the RC was which wouldn't include this?)

The average looks about the same. It seems a bit more peakier on total txn time.

![](https://user-images.githubusercontent.com/558581/187610412-f174c5c6-e6a8-4292-85b8-c15d5d65fd0b.png)

Reply:
Another follow-up, this time I think the changes are actually deployed to `matrix.org` as indicated by that change in the `UP` graph at the top (and `/versions` shows a hash that includes it)

It looks like it does take significantly more time on average (~100ms vs ~10ms before) but we're doing the query a whole lot more less probably because we fixed up those [mis](https://github.com/matrix-org/synapse/pull/13605)-[uses](https://github.com/matrix-org/synapse/pull/13608).

https://grafana.matrix.org/d/000000012/synapse?orgId=1&from=1662362659334&to=1662535459334&viewPanel=180

![](https://user-images.githubusercontent.com/558581/188814736-caaa1a4b-960c-4ac2-98b0-66835d7edbd8.png)

Reply:
This probably has a bigger penalty for other usage patterns, see https://github.com/matrix-org/synapse/issues/13942

Reply:
Probably fixed https://github.com/matrix-org/synapse/issues/13942 via https://github.com/matrix-org/synapse/pull/13958 (another `get_users_in_room` mis-use)

Fixed up another mis-use in https://github.com/matrix-org/synapse/pull/13960

And found another mis-use that's tracked by https://github.com/matrix-org/synapse/issues/13967. Since this involves more than just switching it out, I'm not going to tackle. Feel free to pick it up!

Reply:
This ended up as two functions `get_current_hosts_in_room` and `get_current_hosts_in_room_ordered` so that we don't have to take the performance hit of the extra join in `get_users_in_room`, see https://github.com/matrix-org/synapse/pull/13972